lib/iris/analysis/maths.py

```diff
@@ -311,11 +319,35 @@ def multiply(cube, other, dim=None, in_place=False):
     _assert_is_cube(cube)
     other_unit = getattr(other, 'units', '1')
     new_unit = cube.units * other_unit
-    op = operator.imul if in_place else operator.mul
+    if in_place:
+        _inplace_common(cube, other, 'multiplication')
+        op = operator.imul
+    else:
+        op = operator.mul
     return _binary_op_common(op, 'multiply', cube, other, new_unit, dim,
                              in_place=in_place)
 
 
+def _inplace_common(cube, other, math_op):
+    """
+    Check whether an inplace math operation can take place between `cube` and
+    `other`. It cannot if `cube` has integer data and `other` has float data
+    as the operation will always produce float data that cannot be 'safely'
+    cast back to the integer data of `cube`.
+
+    """
+    if cube.dtype.kind == 'i':
```

Comment:
I'll go as far as unsigned ints but I don't believe there's a use-case for bool checks here too. In the spirit of maximising work not done I'm not going to touch bools. If someone finds there's a problem for having not done so then we can think about it further then...